### PR TITLE
Clean up and optimize OpenTask / read_index

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister

### CODE REVIEW

File: src/librustc/dep_graph/graph.rs [hunk 1]
```diff
@@ -409,7 +418,7 @@ impl DepGraph {
     #[inline]
     pub fn read_index(&self, dep_node_index: DepNodeIndex) {
         if let Some(ref data) = self.data {
-            data.current.borrow_mut().read_index(dep_node_index);
+            data.read_index(dep_node_index);
```
Comment (Zoxc) [inline src/librustc/dep_graph/graph.rs]: Note: We no longer need the global dep graph lock here for parallel queries.

**Review by michaelwoerister — APPROVED** [review]
Looks great, thanks @Zoxc! 

I wonder if we can get rid of the lock around the read-set too at some point. Maybe collect reads per thread, without locking, and merge them during task completion. But correctness is a concern here. We rely on the order of reads during `try_mark_green`. Anyway, no need to solve this now.

@bors r+

**Comment by phansch** [comment]
@michaelwoerister I think bors doesn't work with commands in review comments

**Pull request merged**